File: scripts/wz_img_metadata.py
"""Read unencrypted KMS IMG property metadata; Canvas pixel data is skipped.

String blocks follow the MapleStory.io PKG1 WZReader format. This deliberately
rejects encrypted/unknown formats rather than inventing frame origins or delays.
"""

import struct
from typing import Any


class ImgReader:
    def __init__(self, data: bytes):
        self.data = data
        self.position = 0

    def read(self, length: int) -> bytes:
        if length < 0 or self.position + length > len(self.data):
            raise ValueError("IMG property exceeds file bounds")
        result = self.data[self.position:self.position + length]
        self.position += length
        return result

    def unpack(self, format: str) -> Any:
        return struct.unpack("<" + format, self.read(struct.calcsize("<" + format)))[0]

    def integer(self, format: str = "i") -> int:
        value = self.unpack("b")
        return self.unpack(format) if value == -128 else value

    def string(self) -> str:
        length = self.unpack("b")
        if length <= 0:
            length = self.unpack("i") if length == -128 else -length
            return bytes(
                value ^ ((0xAA + index) & 0xFF)
                for index, value in enumerate(self.read(length))
            ).decode("latin1")
        length = self.unpack("i") if length == 127 else length
        encoded = self.read(length * 2)
        decoded = bytearray()
        for index in range(length):
            value = struct.unpack_from("<H", encoded, index * 2)[0]
            decoded.extend(struct.pack("<H", value ^ ((0xAAAA + index) & 0xFFFF)))
        return decoded.decode("utf-16le")

    def string_block(self) -> str:
        marker = self.unpack("B")
        if marker in (0, 0x73):
            return self.string()
        if marker in (1, 0x1B):
            offset = self.unpack("i")
            if not 0 <= offset < len(self.data):
                raise ValueError("Invalid IMG string offset")
            saved = self.position
            self.position = offset
            result = self.string()
            self.position = saved
            return result
        raise ValueError(f"Unsupported IMG string block: {marker}")

    def properties(self) -> dict[str, Any]:
        self.read(2)
        count = self.integer()
        if count < 0:
            raise ValueError("Invalid IMG property count")
        properties: dict[str, Any] = {}
        for _ in range(count):
            name = self.string_block()
            kind = self.unpack("B")
            value = None
            children = {}
            if kind == 0:
                pass
            elif kind in (2, 11):
                value = self.unpack("h")
            elif kind == 18:
                value = self.unpack("H")
            elif kind in (3, 19):
                value = self.integer()
            elif kind == 20:
                value = self.integer("q")
            elif kind == 4:
                value = self.unpack("f") if self.unpack("B") == 128 else 0
            elif kind == 5:
                value = self.unpack("d")
            elif kind == 8:
                value = self.string_block()
            elif kind == 9:
                length = self.unpack("I")
                end = self.position + length
                if end > len(self.data):
                    raise ValueError("IMG extended property exceeds file bounds")
                kind = self.string_block()
                if kind == "Property":
                    children = self.properties()
                elif kind == "Canvas":
                    self.read(1)
                    if self.unpack("B"):
                        children = self.properties()
                elif kind == "Shape2D#Vector2D":
                    value = {"x": self.integer(), "y": self.integer()}
                elif kind == "UOL":
                    self.read(1)
                    value = self.string_block()
                else:
                    raise ValueError(f"Unsupported IMG extended property: {kind}")
                if self.position > end:
                    raise ValueError("IMG extended property length mismatch")
                self.position = end
            else:
                raise ValueError(f"Unsupported IMG property: {kind} ({name})")
            properties[name] = {"type": kind, "value": value, "children": children}
        return properties


def read_img_metadata(data: bytes) -> dict[str, Any]:
    reader = ImgReader(data)
    if reader.string_block() != "Property":
        raise ValueError("Expected an unencrypted KMS Property IMG")
    result = reader.properties()
    if reader.position != len(data):
        raise ValueError("Unexpected trailing IMG data")
    return result


def find_img_node(root: dict[str, Any], path: str) -> dict[str, Any]:
    node = {"children": root}
    for part in path.split("/"):
        node = node["children"][part]
    return node

File: src/utils/wzImgMetadata.ts
export type ImgNode = {
  type: string | number
  value?: unknown
  children: ImgTree
}
export type ImgTree = Record<string, ImgNode>

// MapleStory.io raw IMG exports use unencrypted PKG1 property strings.
// Canvas pixels and collision/audio payloads are skipped; pixels come from /wz/img.
export const readWzImgMetadata = (buffer: ArrayBuffer): ImgTree => {
  const bytes = new Uint8Array(buffer)
  const view = new DataView(buffer)
  let cursor = 0
  let nodeCount = 0
  const take = (size: number) => {
    if (!Number.isSafeInteger(size) || size < 0 || cursor + size > bytes.length)
      throw new Error('Invalid IMG property bounds')
    const start = cursor
    cursor += size
    return start
  }
  const u8 = () => view.getUint8(take(1))
  const i8 = () => view.getInt8(take(1))
  const i32 = () => view.getInt32(take(4), true)
  const integer = () => {
    const n = i8()
    return n === -128 ? i32() : n
  }
  const string = () => {
    const marker = i8()
    const unicode = marker > 0
    const length = marker === -128 || marker === 127 ? i32() : Math.abs(marker)
    const start = take(length * (unicode ? 2 : 1))
    let result = ''
    for (let i = 0; i < length; i++)
      result += String.fromCharCode(
        unicode
          ? view.getUint16(start + i * 2, true) ^ ((0xaaaa + i) & 0xffff)
          : bytes[start + i] ^ ((0xaa + i) & 0xff)
      )
    return result
  }
  const stringBlock = () => {
    const marker = u8()
    if (marker === 0 || marker === 0x73) return string()
    if (marker !== 1 && marker !== 0x1b)
      throw new Error(`Unsupported IMG string: ${marker}`)
    const offset = i32()
    if (offset < 0 || offset >= bytes.length)
      throw new Error('Invalid IMG string offset')
    const saved = cursor
    cursor = offset
    const result = string()
    cursor = saved
    return result
  }
  const properties = (depth = 0): ImgTree => {
    if (depth > 64) throw new Error('IMG property nesting limit exceeded')
    take(2)
    const count = integer()
    if (count < 0 || (nodeCount += count) > 500000)
      throw new Error('Invalid IMG property count')
    const result: ImgTree = Object.create(null)
    for (let i = 0; i < count; i++) {
      const name = stringBlock()
      let type: string | number = u8()
      let value: unknown
      let children: ImgTree = Object.create(null)
      switch (type) {
        case 0:
          break
        case 2:
        case 11:
          value = view.getInt16(take(2), true)
          break
        case 18:
          value = view.getUint16(take(2), true)
          break
        case 3:
        case 19:
          value = integer()
          break
        case 20: {
          const n = i8()
          value = n === -128 ? Number(view.getBigInt64(take(8), true)) : n
          break
        }
        case 4:
          value = u8() === 128 ? view.getFloat32(take(4), true) : 0
          break
        case 5:
          value = view.getFloat64(take(8), true)
          break
        case 8:
          value = stringBlock()
          break
        case 9: {
          const size = view.getUint32(take(4), true)
          const end = cursor + size
          if (end > bytes.length)
            throw new Error('Invalid IMG extended property bounds')
          type = stringBlock()
          if (type === 'Property') children = properties(depth + 1)
          else if (type === 'Canvas') {
            take(1)
            if (u8()) children = properties(depth + 1)
          } else if (type === 'Shape2D#Vector2D')
            value = { x: integer(), y: integer() }
          else if (type === 'UOL') {
            take(1)
            value = stringBlock()
          } else if (
            !['Shape2D#Convex2D', 'Sound_DX8', 'RawData', 'Video'].includes(
              type
            )
          )
            throw new Error(`Unsupported IMG property: ${type}`)
          if (cursor > end) throw new Error('IMG property length mismatch')
          cursor = end
          break
        }
        default:
          throw new Error(`Unsupported IMG type: ${type}`)
      }
      result[name] = { type, value, children }
    }
    return result
  }
  if (stringBlock() !== 'Property')
    throw new Error('Unsupported/encrypted IMG export')
  const root = properties()
  if (bytes.subarray(cursor).some((byte) => byte !== 0))
    throw new Error('Unexpected IMG trailing data')
  return root
}

export const normalizeImgPath = (path: string) => {
  const parts: string[] = []
  for (const part of path.split('/')) {
    if (part === '..') {
      if (!parts.length) throw new Error('IMG link leaves its root')
      parts.pop()
    } else if (part && part !== '.') parts.push(part)
  }
  return parts.join('/')
}

export const findWzImgNode = (tree: ImgTree, path: string): ImgNode => {
  let node: ImgNode = { type: 'Property', children: tree }
  for (const part of normalizeImgPath(path).split('/').filter(Boolean)) {
    const next = node.children[part]
    if (!next) throw new Error(`Missing IMG node: ${path}`)
    node = next
  }
  return node
}
